### PR TITLE
cpu/efm32: add RTT_FREQUENCY support to efm32

### PR DESCRIPTION
**Pull request by kfessel**
### Contribution description

make efm32 rtt frequency configurable by setting RTT_FREQUENCY

### Testing procedure

**I can not test that, I don't have such a BOARD**  see **Test results** :arrow_down:

```
tests/ztimer_xsec$ BOARD=<some efm32> USEMODULE=ztimer_periph_rtt CFLAGS=-DRTT_FREQEUNCY=32768 make flash test
```
expected to print SUCCESS and some time information 
(start Time; SEC ~1*10^^6 later; MSEC ~200*10^^3 later; USEC ~100*10^^3 later)
```
tests/periph_rtt$ BOARD=<some efm32> USEMODULE=ztimer_periph_rtt CFLAGS=-DRTT_FREQUENCY=32768 make flash term
```
expected to print Hello every 5 seconds 

RTT_FREQUENCY may be chosen from 2^^0 to 2^^15  other values should fail at precompile-time 

please test with 
```
1024 (2^^10)  
    periph_rtt should print the line "Setting initial alarm to now + 5 s (5ddd) (d any decimal)"
4096 (2^^12)
    periph_rtt should print the line "Setting initial alarm to now + 5 s (20ddd) (d any decimal)"
32768 (2^^15)
    periph_rtt should print the line "Setting initial alarm to now + 5 s (16dddd) (d any decimal)"
``` 
periph_rtt  also prints `RTT_FREQUENCY: ...` after reset if possible please verify

### Test results

The first testing procedure was wrong: 
    I wrongly assumend  an enviroment RTT_FREQUENCY is translated to a define RTT_FREQUENCY somewhere in the build process

many thanks to @chrysn

TEST where successfully redone by @chrysn :+1: 
stk3700 and sltb001a (one of them is a series0 the other series1 one has RTC the other RTCC) -> both code paths are checked
messages where checked (numbers and success)

### Issues/PRs references

https://forum.riot-os.org/t/default-rtt-frequency/3160/4

### CODE REVIEW

**Comment by chrysn** [comment]
I don't understand the components, but have the stk3700 to test it:

<details><summary><code>$ BOARD=stk3700 USEMODULE=ztimer_periph_rtt RTT_FREQUENCY=4096 make -C tests/ztimer_xsec all flash term PROGRAMMER=openocd</code></summary>
<code><pre>
[...]
2021-04-16 18:11:33,344 # START
2021-04-16 18:11:33,345 # main(): This is RIOT! (Version: 2021.04-devel-1355-gf1a58-p-efm32-rtt-freq)
2021-04-16 18:11:33,345 # starting ztimers
2021-04-16 18:11:33,345 # time Wait:    3947036
2021-04-16 18:11:33,365 # waiting for locks
2021-04-16 18:11:33,462 # USEC
2021-04-16 18:11:33,562 # MSEC
2021-04-16 18:11:35,420 # SEC
2021-04-16 18:11:35,420 # time SEC:     5999812
2021-04-16 18:11:35,421 # time MSEC:    4147028
2021-04-16 18:11:35,421 # time USEC:    4047044
2021-04-16 18:11:35,421 # SUCCESS!
</pre></code>
</details>

<details><summary><code>$ BOARD=stk3700 USEMODULE=ztimer_periph_rtt RTT_FREQUENCY=32768 make -C tests/ztimer_xsec all flash term PROGRAMMER=openocd</code></summary>
<code><pre>
[...]
2021-04-16 18:19:17,672 # START
2021-04-16 18:19:17,675 # main(): This is RIOT! (Version: 2021.04-devel-1355-gf1a58-p-efm32-rtt-freq)
2021-04-16 18:19:17,676 # starting ztimers
2021-04-16 18:19:17,677 # time Wait:    8173464
2021-04-16 18:19:17,693 # waiting for locks
2021-04-16 18:19:17,790 # USEC
2021-04-16 18:19:17,890 # MSEC
2021-04-16 18:19:20,522 # SEC
2021-04-16 18:19:20,522 # time SEC:     10999700
2021-04-16 18:19:20,523 # time MSEC:    8373456
2021-04-16 18:19:20,523 # time USEC:    8273472
2021-04-16 18:19:20,524 # SUCCESS!
</pre></code>
</details>

<details><summary><code>$ BOARD=stk3700 USEMODULE=ztimer_periph_rtt RTT_FREQUENCY=1 make -C tests/ztimer_xsec all flash term PROGRAMMER=openocd</code></summary>
<code><pre>
[...]
2021-04-16 18:20:14,680 # START
2021-04-16 18:20:14,681 # main(): This is RIOT! (Version: 2021.04-devel-1355-gf1a58-p-efm32-rtt-freq)
2021-04-16 18:20:14,681 # starting ztimers
2021-04-16 18:20:14,682 # time Wait:    2428796
2021-04-16 18:20:14,701 # waiting for locks
2021-04-16 18:20:14,798 # USEC
2021-04-16 18:20:14,898 # MSEC
2021-04-16 18:20:17,276 # SEC
2021-04-16 18:20:17,277 # time SEC:     4999836
2021-04-16 18:20:17,278 # time MSEC:    2628784
2021-04-16 18:20:17,279 # time USEC:    2528804
2021-04-16 18:20:17,280 # SUCCESS!
</pre></code>
</details>

Hope this helps.

**Comment by chrysn** [comment]
Ran tests/periph_rtt with the same board and RTT_FREQUENCY values 1, 4096 and 32768; all gave a Hello every 5 seconds.

**Comment by chrysn** [comment]
Running the same tests on sltb001a for series1 coverage:

<details><summary><code>BOARD=sltb001a USEMODULE=ztimer_periph_rtt RTT_FREQUENCY=32768 make -C tests/ztimer_xsec all compile-commands flash term PROGRAMMER=openocd</code></summary>
<code><pre>
2021-04-16 18:50:03,814 # START
2021-04-16 18:50:03,815 # main(): This is RIOT! (Version: 2021.04-devel-1355-gf1a58-p-efm32-rtt-freq)
2021-04-16 18:50:03,816 # starting ztimers
2021-04-16 18:50:03,816 # time Wait:    3679888
2021-04-16 18:50:03,817 # waiting for locks
2021-04-16 18:50:03,914 # USEC
2021-04-16 18:50:04,013 # MSEC
2021-04-16 18:50:05,154 # SEC
2021-04-16 18:50:05,154 # time SEC:     5018680
2021-04-16 18:50:05,154 # time MSEC:    3879884
2021-04-16 18:50:05,155 # time USEC:    3779896
2021-04-16 18:50:05,155 # SUCCESS!
</pre></code>
</details>


<details><summary><code>BOARD=sltb001a USEMODULE=ztimer_periph_rtt RTT_FREQUENCY=1 make -C tests/ztimer_xsec all compile-commands flash term PROGRAMMER=openocd</code></summary>
<code><pre>
2021-04-16 18:51:23,951 # START
2021-04-16 18:51:23,952 # main(): This is RIOT! (Version: 2021.04-devel-1355-gf1a58-p-efm32-rtt-freq)
2021-04-16 18:51:23,952 # starting ztimers
2021-04-16 18:51:23,952 # time Wait:    9486632
2021-04-16 18:51:23,954 # waiting for locks
2021-04-16 18:51:24,051 # USEC
2021-04-16 18:51:24,150 # MSEC
2021-04-16 18:51:25,505 # SEC
2021-04-16 18:51:25,505 # time SEC:     11041040
2021-04-16 18:51:25,505 # time MSEC:    9686628
2021-04-16 18:51:25,505 # time USEC:    9586640
2021-04-16 18:51:25,505 # SUCCESS!
</pre></code>
</details>

and `BOARD=sltb001a USEMODULE=ztimer_periph_rtt RTT_FREQUENCY=1 make -C tests/periph_rtt all compile-commands flash term PROGRAMMER=openocd` gave me a Hello every 5 seconds; same with =32768.

**Comment by benpicco** [comment]
So all tested, the change is simple too - want to give this an ACK?

**Review by chrysn — APPROVED** [review]
I was about to say that I'd have to dig into the timer systems again, but now looking at the code, I can see how it *is* a simple change. Don't see precisely where which timers are selected, but I could verify that the code paths are active in the tests, and everything else checks out, so here we go.

**Review by fjmolinas — CHANGES_REQUESTED** [review]
How you mind adding headers for the configuration like done for other platforms:

```
/**
 * @name    Real time counter configuration
 * @{
 */
#define RTT_MAX_VALUE       (0x00ffffff)         /* 24bit */
#define RTT_MAX_FREQUENCY   (32768U)             /* in Hz */
#define RTT_MIN_FREQUENCY   (8U)                 /* in Hz */
#define RTT_CLOCK_FREQUENCY (32768U)             /* in Hz, LFCLK*/

#ifndef RTT_FREQUENCY
#define RTT_FREQUENCY       (1024U)              /* in Hz */
#endif
/** @} */
```

**Comment by fjmolinas** [comment]
> How you mind adding headers for the configuration like done for other platforms:
> 
> ```
> /**
>  * @name    Real time counter configuration
>  * @{
>  */
> #define RTT_MAX_VALUE       (0x00ffffff)         /* 24bit */
> #define RTT_MAX_FREQUENCY   (32768U)             /* in Hz */
> #define RTT_MIN_FREQUENCY   (8U)                 /* in Hz */
> #define RTT_CLOCK_FREQUENCY (32768U)             /* in Hz, LFCLK*/
> 
> #ifndef RTT_FREQUENCY
> #define RTT_FREQUENCY       (1024U)              /* in Hz */
> #endif
> /** @} */
> ```

CUrrently I think all `efm32` platforms are defininf:

```
/**
 * @name    RTT configuration
 * @{
 */
#define RTT_MAX_VALUE       (0xFFFFFF)
#define RTT_FREQUENCY       (1U)
/** @} */
```

This could all be placed in one common place, and without header guards the frequency isn't really configurable yet, or am I missign something?

**Comment by kfessel** [comment]
@fjmolinas you are right; I was assuming there is a CFLAG=-DRTT_FREQUNCY=$(RTT_FREQUENCY) in the RIOT makesystem

**Comment by chrysn** [comment]
Sorry for not spotting this, should have been obvious :-(

File: cpu/efm32/include/periph_cpu.h [hunk 1]
```diff
@@ -119,6 +120,18 @@ typedef struct {
 } dac_chan_conf_t;
 #endif
 
+/**
+ * @name    Real time counter configuration
+ * @{
+ */
+#define RTT_MAX_VALUE       _RTC_CNT_MASK        /* mask has all bits set ==> MAX*/
```
Comment (chrysn) [inline cpu/efm32/include/periph_cpu.h:127]: `BOARD=sltb001a make -C tests/periph_rtt all ` does not run through for me:

```
"make" -C /home/chrysn/git/RIOT-doc-fixes/build/pkg/gecko_sdk/dist/emlib/src
In file included from /home/chrysn/git/RIOT-doc-fixes/boards/sltb001a/include/periph_conf.h:24,
                 from /home/chrysn/git/RIOT-doc-fixes/tests/periph_rtt/main.c:29:
/home/chrysn/git/RIOT-doc-fixes/tests/periph_rtt/main.c: In function 'cb':
/home/chrysn/git/RIOT-doc-fixes/cpu/efm32/include/periph_cpu.h:127:29: error: '_RTC_CNT_MASK' undeclared (first use in this function); did you mean 'RTCC_BASE'?
 #define RTT_MAX_VALUE       _RTC_CNT_MASK        /* mask has all bits set ==> MAX*/
                             ^~~~~~~~~~~~~
/home/chrysn/git/RIOT-doc-fixes/tests/periph_rtt/main.c:43:13: note: in expansion of macro 'RTT_MAX_VALUE'
     last &= RTT_MAX_VALUE;
```

Either something is wrong with the include sequence, or that mask is only defined for some of the gecko devices.

**Review by chrysn — CHANGES_REQUESTED** [review]

File: cpu/efm32/periph/rtt_series0.c [hunk 2]
```diff
@@ -34,10 +34,48 @@ typedef struct {
 
 static rtt_state_t rtt_state;
 
+/* prescaler of 32768 = 1 s of resolution and overflow each 194 days */
+#if RTT_FREQUENCY == 1
+#define cmuClkDiv cmuClkDiv_32768
```
[the remark this answers is not in the document]
Comment (fjmolinas) [reply cpu/efm32/periph/rtt_series0.c:39]: Why change this it was already set to `RTT_MAX_VALUE       (0xFFFFFF)` for all boards, was that value not correct?

File: cpu/efm32/include/periph_cpu.h (hunk 1 above)
Comment (kfessel) [reply cpu/efm32/include/periph_cpu.h:127]: seems like some call it _RTCC_CNT_MASK (maybe the same as the series 1/0 split)

File: cpu/efm32/periph/rtt_series0.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (kfessel) [reply cpu/efm32/periph/rtt_series0.c:39]: took the inspiration from the emlib define name 
I change

[the remark this answers is not in the document]
Comment (fjmolinas) [reply cpu/efm32/periph/rtt_series0.c:39]: > Why change this it was already set to `(0xFFFFFFFF)` for all boards, was that value not correct?

ups seems I commented on the wrong line, this was for `RTT_MAX_VALUE`

[the remark this answers is not in the document]
Comment (kfessel) [reply cpu/efm32/periph/rtt_series0.c:39]: some BOARDS have 24Bit some 32Bit (maybe the same as the RTC vs RTCC or series0 vs series1 cut)

File: cpu/efm32/include/periph_cpu.h (hunk 1 above)
Comment (fjmolinas) [reply cpu/efm32/include/periph_cpu.h:127]: Seems like `_RTC_CNT_MASK ` is always 24 bit

Comment (fjmolinas) [reply cpu/efm32/include/periph_cpu.h:127]: YOu could just do:

```
#ifdef _RTC_CNT_MASK 
#define RTT_MAX_VALUE       _RTC_CNT_MASK
#endif
#ifdef _RTCC_CNT_MASK 
#define RTT_MAX_VALUE       _RTCC_CNT_MASK 
#endif
```

If there is still another one the compiler will catch it.

Comment (kfessel) [reply cpu/efm32/include/periph_cpu.h:129]: some BOARDS have 24Bit some 32Bit for RTT_MAX_VALUE (maybe the same as the RTC vs RTCC or series0 vs series1 cut)

**Review by fjmolinas — APPROVED** [review]
ACK, my changes have been addressed, trusting @chrysn regarding testing.

**Review by chrysn — APPROVED** [review]
The testing procedure has a `CFLAG` paramter rather than a `CFLAGS` and a typo in the FREQUNCY, which upset my initial tests (merely a note to any other testers).

In test/periph_rtt, RTT_FREQUENCY gets shown right, the initial alarm times are in the right order of magnitude, the hellos show up and the ztimer_xsec show "SUCCESS" too. (With selected values of frequency at stk3700 and sltb001a).

**Comment by kfessel** [comment]
Added your test results to the descricption and and fixed the test procedure typos

**Comment by kfessel** [comment]
squashed the fixups

**Comment by kfessel** [comment]
seems like this is ready to merge

**Pull request merged**